src/test/jtx/batch.h

```diff
@@ -0,0 +1,185 @@
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_TEST_JTX_BATCH_H_INCLUDED
+#define RIPPLE_TEST_JTX_BATCH_H_INCLUDED
+
+#include <test/jtx/Account.h>
+#include <test/jtx/Env.h>
+#include <test/jtx/amount.h>
+#include <test/jtx/owners.h>
+#include <test/jtx/tags.h>
+#include <concepts>
+#include <cstdint>
+#include <optional>
+
+namespace ripple {
+namespace test {
+namespace jtx {
+
+/** Batch operations */
+namespace batch {
+
+/** Batch. */
+Json::Value
+batch(
+    jtx::Account const& account,
+    uint32_t seq,
+    STAmount const& fee,
+    std::uint32_t flags);
+
+/** Adds a new Batch Txn on a JTx. */
+class add
+{
+private:
+    Json::Value txn_;
+    std::uint32_t seq_;
+    std::optional<std::uint32_t> ticket_;
+
+public:
+    add(Json::Value const& txn,
+        std::uint32_t const& sequence,
+        std::optional<std::uint32_t> const& ticket = std::nullopt,
+        std::optional<std::uint32_t> const& fee = std::nullopt)
+        : txn_(txn), seq_(sequence), ticket_(ticket)
+    {
+    }
+
+    void
+    operator()(Env&, JTx& jtx) const;
+};
+
+/** Set a batch signature on a JTx. */
+class sig
+{
+public:
+    struct Reg
+    {
+        Account acct;
+        Account sig;
+
+        Reg(Account const& masterSig) : acct(masterSig), sig(masterSig)
+        {
+        }
+
+        Reg(Account const& acct_, Account const& regularSig)
+            : acct(acct_), sig(regularSig)
+        {
+        }
+
+        Reg(char const* masterSig) : acct(masterSig), sig(masterSig)
+        {
+        }
+
+        Reg(char const* acct_, char const* regularSig)
+            : acct(acct_), sig(regularSig)
+        {
+        }
+
+        bool
+        operator<(Reg const& rhs) const
+        {
+            return acct < rhs.acct;
+        }
+    };
+
+    std::vector<Reg> signers;
+
+public:
+    sig(std::vector<Reg> signers_);
+
+    template <class AccountType, class... Accounts>
+        requires std::convertible_to<AccountType, Reg>
+    explicit sig(AccountType&& a0, Accounts&&... aN)
+        : sig{std::vector<Reg>{
+              std::forward<AccountType>(a0),
+              std::forward<Accounts>(aN)...}}
+    {
+    }
+
+    void
+    operator()(Env&, JTx& jt) const;
+};
+
+/** Set a batch multi signature on a JTx. */
+class msig
+{
+public:
+    struct Reg
```

Comment:
There doesn't look to be a difference between this `Reg` and the one above it. Why not make it a struct outside of the classes, to avoid repetition?

Reply:
refactored msig: https://github.com/XRPLF/rippled/pull/5060/commits/ef49db7c5e177747296b02b653c04b60dbfc5518